apisix/plugins/ai-rag.lua

```diff
@@ -0,0 +1,168 @@
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local http = require("resty.http")
+local ngx_req = ngx.req
+local core = require("apisix.core")
+local decorate = require("apisix.plugins.ai-prompt-decorator").__decorate
+local next = next
+local require = require
+
+local azure_ai_search_schema = {
+    type = "object",
+    properties = {
+        endpoint = {
+            type = "string",
+        },
+        api_key = {
+            type = "string",
+        },
+    }
+}
+
+local azure_openai_embeddings = {
+    type = "object",
+    properties = {
+        endpoint = {
+            type = "string",
+        },
+        api_key = {
+            type = "string",
+        },
+    },
+    required = { "endpoint", "api_key" }
+}
+
+
+local schema = {
+    type = "object",
+    properties = {
+        type = "object",
+        embeddings_provider = {
+            type = "object",
+            properties = {
+                azure_openai = azure_openai_embeddings
+            },
+            -- change to enum while implementing support for other search services
+            required = { "azure_openai" },
+        },
+        vector_search_provider = {
+            type = "object",
+            properties = {
+                azure_ai_search = azure_ai_search_schema
+            },
+            -- change to enum while implementing support for other search services
+            required = { "azure_ai_search" }
+        },
+    },
+    required = { "embeddings_provider", "vector_search_provider" }
+}
+
+local request_schema = {
+    type = "object",
+    properties = {
+        ai_rag = {
+            type = "object",
+            properties = {
+                vector_search = {},
+                embeddings = {},
+            },
+            required = { "vector_search", "embeddings" }
+        }
+    }
+}
+
+local _M = {
+    version = 0.1,
+    priority = 1060, -- TODO check with other ai plugins
+    name = "ai-rag",
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+function _M.access(conf, ctx)
+    local httpc = http.new()
+    local body_tab, err = core.request.get_json_request_body_table()
+    if not body_tab then
+        return 400, err
+    end
+    if not body_tab["ai_rag"] then
+        core.log.error("request body must have \"ai-rag\" field")
+        return 400
+    end
+
+    local embeddings_provider = next(conf.embeddings_provider)
```

Comment:
what if `embeddings_provider` have multiple properties?

Reply:
the schema should ensure only one provider can be configured.